File: 2226_maximum_candies_allocated_to_k_children.cpp
#include <vector>
#include <numeric>
using namespace std;

class Solution {
public:
    int maximumCandies(vector<int>& a, long long k) {
        long long sum = accumulate(a.begin(), a.end(), 0LL);
        if (sum < k) return 0;
        int l = 1, r = sum / k;
        while (l < r) {
            int m = (l + r + 1) / 2;
            if (accumulate(a.begin(), a.end(), 0LL, [&](long long s, int x) { return s + x / m; }) < k)
                r = m - 1;
            else
                l = m;
        }
        return l;
    }
};

File: 2226_maximum_candies_allocated_to_k_children.py
class Solution(object):
    def maximumCandies(self, candies, k):
        if sum(candies) < k:
            return 0
        l, r = 1, sum(candies) // k
        while l < r:
            m = (l + r + 1) // 2
            if sum(x // m for x in candies) < k:
                r = m - 1
            else:
                l = m
        return l
